### ISSUE TITLE
Avoid deleting/reallocating z/OS datasets

### ISSUE THREAD

**Issue by andrew890** [wizardofzos/zigi#375]
As a Systems Programmer, source control seems like an ideal way to track changes to system configuration e.g. SYS1.PARMLIB, SYS1.PROCLIB etc.

In fact since Rocket Git became available I do have a process to copy members from these datasets to OMVS and track the changes in Git, similar to what Zigi does.

Zigi seemed like it might be a useful addition to this process. However I discovered that Zigi will delete and reallocate datasets during various processes. This is obviously unacceptable for these datasets, and I suspect the same applies to many other datasets where Git tracking would be useful.

I would like to see Zigi use safer processes for managing z/OS datasets.
- never delete and reallocate a dataset without confirmation
- where possible, work with individual members rather than a wholesale copy of members to the dataset

```
git diff --name-status branch1 branch2
```
may be useful to work out what really needs to be copied rather than copying everything, with the advantage that checkouts with minimal changes could also be much faster.

**Comment by lbdyck** [comment]
ZIGI does a reallocation - true
ZIGI in many cases only copies from OMVS to z/OS the changed set of members - true

Can you identify the instances where ZIGI is doing the reallocation that you do not want it to - that will help us.

**Comment by andrew890** [comment]
For many of the datasets I would like to manage i.e. system configuration (SYS1.PARMLIB, SYS1.PROCLIB etc.) you would never want it deleted/reallocated. There may also be rules about whether they can be PDSE.

I could also see value in managing members in e.g. ISPF libraries that are always allocated so you can replace members but not delete the dataset.

You can of course create another copy of the datasets for Zigi management and manually copy back and forth but that just shifts the manual processing step that Zigi is trying to eliminate.

**Comment by wizardofzos** [comment]
Chiming in :)

I'll admin I broke my system once by adding the PARMLIB dataset to zigi and then branching and pulling causing the 'delete define' of the PARMLIB. And yes on another volume.

As Lionel has stated, zigi 'emulates' GIT as much as possible. Taking the USS bit of the REPO as 'the truth' and making sure the z/OS-datasets reflect that truth.

The earlier version of zigi used to be much more 'agressive' when it came to deleting and allocating the 'target' (z/OS) datasets. the current version tries to to that as little as possible. 

As stated I have the same issues with the PARMLIB (and the PROCLIB too)

Maybe it's an option to add some kind of 'flagging' option when adding a dataset to state you never want this dataset to be deleted when refreshing the repo?

On the other hand, I've 'worked' around this by having a set of 'source libraries' (that are zigi/git managed) that get 'deployed' to the actual PARM- and PROCLIBS....

That REPO (currently) has 3 'folders' PARMLIB, PROCLIB and DEPLOY and the DEPLOY library has jobs to copy the contents from PAMLIB/PROCLIB to the 'real' datasets...

**Comment by andrew890** [comment]
I would disagree that Git is 'the truth'. The z/OS datasets are 'the truth'. What is in Git is a record of specific parts of the truth at a certain time.

There are lots of datasets on z/OS systems that would be useful to track in git, so I like the idea of Zigi. However, many of them cannot be arbitrarily delete/defined. I think the current implementation will cause a lot of people to have accidents like your PARMLIB example.

**Comment by lbdyck** [comment]
I've been mulling this over and some thoughts:

1. Git is a repository for development and should not be used for active production version data - only copies of production in the same way that SMP/E target datasets should never be the actual production libraries.
2. Take a look at the ZGBATCH exec included with ZIGI 3.0 and see if that works for what you are asking - a copy z/OS to OMVS only and never the reverse.
   - this does not preclude someone from going into ZIGI and initiating some action that would update the z/OS datasets so caution is mandated

**Comment by andrew890** [comment]
My thoughts: Git is a version control system, and is potentially useful anytime you want to keep a history of what was modified, when, why and by whom. Limiting it to development work seems unnecessarily restrictive. Tracking e.g. modifications to SYS1.PARMLIB seems like a perfect use case.

(Cloning a SMP/E target zone for e.g. every SYSRES is actually quite useful. Applying maintenance to production libraries is not a good idea, but it is nice to be able to query the history available in SMP/E for a specific set of production libraries, compare maintenance levels between different SYSRES etc.)

ZGBATCH - why is it necessary to specify HLQ and QUALIGNR? I think that should be part of the metadata in the working copy. It seems like allowing different people to use the same git working copy with different z/OS datasets is not a good idea.

**Comment by lbdyck** [comment]
ZGBATCH requires the hlq/qualignr as that information is stored in an ISPF table in the users table/profile dataset and thus may not be accessible to a different user. The info could be stored in the .zigi/dsn file but it isn't at this point.

As for allowing different people to share the same git working copy the assumption is the repository would be the same z/OS datasets and same repository for both. Something that is shared for managing system level datasets or some other set of datasets that are only to be managed with the expectation that the OMVS files would never be used to replace the z/OS datasets (although that can't be prevented at this point).

**Comment by andrew890** [comment]
Hlq/qualignr need to be working copy specific. They need to be stored in the working copy but not committed, a file in the .git directory might be most appropriate. I might clone one repository multiple times under different HLQs.

It seems like it would be bad to have different people, or ZGBATCH use different values for hlq/qualignr for the same repository. That is why I think it needs to be part of the working copy metadata.

**Comment by lbdyck** [comment]
At this point the hlq/qualignr is in the users ISPF table.  To change it would be an effort and for zgbatch to access the ispf table could be problematic depending upon the userid using it and if it is open or not by the ISPF user. I wrote ZGBATCH to demonstrate that it could be done. I'm sure it can be improved upon.

**Comment by lbdyck** [comment]
I think what you are looking for is a repository where the z/OS datasets are in read-only status as far as ZIGI are concerned.

For that I'm thinking:

1. Prevent cloning into existing z/OS datasets (based upon flag in .zigi/dsn file)
2. Prevent z/OS PDS deletes when the OMVS copy of the member is missing                                               
3. Check in ZIGICKOT and ZIGISTAT for additional check
4. Do NOT allow Branch, Merge, STASH, FLOW, Replace
5. Do NOT allow Rename (RN) or Remove (RM)       
6. Should only be a one way pipe - z/OS to OMVS             
7. Flag in .zigi/dsn                                        
   - need special record for this                           
8. ISPF Variable in the zigi repository table               
9. To create the repository use CREATE as today             
   - option on the panel for Read Only                      
10. Option on the ZIGIRSET (SET) popup to set Read Only     
   - do NOT allow it to be removed ??         
11. Do NOT allow History RECOVER
12. Flag somewhere in the Current Repository display (ZIGIREPO panel)
13. Do NOT allow ZIGISTATs to add ISPF statistics if any members are blank/null 
    - means the members will be detected as changed from an ISPF stat perspective and copied to OMVS for Git to determine if a change has been made

What am I missing?

**Comment by andrew890** [comment]
I'm not looking for any "special" processing.

What I would like to see is Zigi to be as consistent as possible with the way git works, and also to avoid doing things that might be unpleasant surprises to the person using it e.g. deleting and reallocating a dataset.

Exactly what that looks like I'm not sure, which is why I suggested getting input form other z/OS people who are familiar with git.

I do think that it should not be possible to have different people using the same working copy to refer to different datasets.

**Comment by lbdyck** [comment]
I believe that ZIGI is as consistent as we can make it to the way Git works. Please let me know where ZIGI fails this.

As for multiple users accessing the same repository - for this read-only use case i don't see an issue with that - please let me know where it fails.

**Comment by lbdyck** [comment]
See the v3r02-dev branch - should help with this.
